### PR TITLE
feat: supports karpenter.sh/instance-price node label

### LINKED ISSUE #1289: [AWS] Attach Spot Pricing Data to Provisioned Nodes
**Tell us about your request**
What do you want us to build?

EC2 Spot Pricing data can be easily accessed via an EC2 API within Karpenter. It would be helpful to attach the spot price at the time of provisioning a node so that the user can have more visibility into a cluster's node cost with Karpenter's extreme flexibility. (This came up in a community meeting from @mikesir87 )

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
What outcome are you trying to achieve, ultimately, and why is it hard/impossible to do right now? What is the impact of not having this problem solved? The more details you can provide, the better we'll be able to understand and solve the problem.

**Are you currently working around this issue?**
How are you currently solving this problem?

Looking up and attributing back to launched nodes. 

**Additional context**
Anything else we should know?

https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-spot-price-history.html

**Attachments**
If you think you might have additional information that you'd like to include via an attachment, please do - we'll take a look. (Remember to remove any personally-identifiable information.)

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### LINKED ISSUE #3237: Label node with instance price when using spot instances
### Tell us about your request

When node labeled with `karpenter.sh/capacity-type: spot` I would like to see a label with price of this spot instance. Something like `karpenter.sh/instance-price: 0.0049`.

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

I'm using Prometheus Operator and for reporting purpose about requested spot instances by karpenter it is required to have information about price of the instance.

### Are you currently working around this issue?

No

### Additional Context

_No response_

### Attachments

_No response_

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by valdisrigdon**
Fixes #1289 and Fixes #3237

**Description**

Adds a new node label to all provisioned nodes with the hourly price, in USD, that was used when provisioning this node.

**How was this change tested?**

* `make test`

**Does this change impact docs?**
- [x] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [ ] No

**Release Note**

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/apis/v1alpha1/register.go [hunk 1]
```diff
@@ -75,6 +75,7 @@ var (
 	LabelInstanceGPUCount                     = LabelDomain + "/instance-gpu-count"
 	LabelInstanceGPUMemory                    = LabelDomain + "/instance-gpu-memory"
 	LabelInstanceAMIID                        = LabelDomain + "/instance-ami-id"
+	LabelInstancePrice                        = LabelDomain + "/instance-price"
```
[the remark this answers is not in the document]
Comment (valdisrigdon) [reply pkg/apis/v1alpha1/register.go:78]: The price is the hourly cost in USD at the time of provisioning. It's a float.

**Comment by bwagner5** [comment]
I chatted with the team about adding this and there were several concerns:

1. This is a pricing estimate at launch which may not be correct in a lot of situations (savings plans, RIs, stale pricing data, etc). 
2. The price can change for spot each hour
3. Adding this as a label may make some user want to use this in their requirements, which I don't think is a great idea. It's also not possible as a float value. 

Would adding the pricing data in the Karpenter logs when the node is provisioned be enough information?

**Comment by valdisrigdon** [comment]
> I chatted with the team about adding this and there were several concerns:
> 
> 1. This is a pricing estimate at launch which may not be correct in a lot of situations (savings plans, RIs, stale pricing data, etc).

I agree that this doesn't take into account those pieces, but this was intended to expose some kind of pricing information to give an estimate on costs, not be a cost accounting system.

> 2. The price can change for spot each hour

I was unaware of this. 

> 3. Adding this as a label may make some user want to use this in their requirements, which I don't think is a great idea. It's also not possible as a float value.

I would agree that using it in the requirements isn't a good idea.

> 
> Would adding the pricing data in the Karpenter logs when the node is provisioned be enough information?

This would be a great start, but we'd love to be able to get this in a prometheus metric so we can do some rough pod cost calculations.

Would renaming the label to `instance-price-estimate` reduce the concerns?

My other idea on this was to expose the `pricing.Provider` on demand/spot pricing as prometheus metrics instead. This way we could use promql joins from node metrics to pricing metrics as karpenter knew at the given time. This would help with issue 2 above where spot prices can change hourly, and with issue 3 where you can't use them as requirements. It would not help with issue 1, but perhaps with clever metrics names, such as including `estimate` it would be clear?

**Comment by bwagner5** [comment]
> My other idea on this was to expose the pricing.Provider on demand/spot pricing as prometheus metrics instead. This way we could use promql joins from node metrics to pricing metrics as karpenter knew at the given time. This would help with issue 2 above where spot prices can change hourly, and with issue 3 where you can't use them as requirements. It would not help with issue 1, but perhaps with clever metrics names, such as including estimate it would be clear?

I'd prefer to go with the prometheus metric and/or including the price in the log when the instance is launched for now.   And yes, I think naming it as price-estimate is enough information. Documentation on the metrics page can also explain why it may not be accurate.

**Comment by valdisrigdon** [comment]
Closing in favor of https://github.com/aws/karpenter/pull/3613

**Pull request closed without merging**